src/base/ERC4626Proxy.sol

```diff
@@ -0,0 +1,201 @@
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            ACCOUNTING LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function totalAssets() public view virtual returns (uint256) {
+        return convertToAssets(underlyingVault.balanceOf(address(this)));
+    }
+
+    function convertToShares(uint256 assets) public view virtual returns (uint256) {
+        return underlyingVault.convertToShares(assets);
+    }
+
+    function convertToAssets(uint256 shares) public view virtual returns (uint256) {
+        return underlyingVault.convertToAssets(shares);
+    }
+
+    function previewDeposit(uint256 assets) public view virtual returns (uint256) {
+        return underlyingVault.previewDeposit(assets);
+    }
+
+    function previewMint(uint256 shares) public view virtual returns (uint256) {
+        return underlyingVault.previewMint(shares);
+    }
+
+    function previewWithdraw(uint256 assets) public view virtual returns (uint256) {
+        return underlyingVault.previewWithdraw(assets);
+    }
+
+    function previewRedeem(uint256 shares) public view virtual returns (uint256) {
+        return underlyingVault.previewRedeem(shares);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                     DEPOSIT/WITHDRAWAL LIMIT LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function maxDeposit(address) public view returns (uint256) {
+        return underlyingVault.maxDeposit(address(this));
+    }
+
+    function maxMint(address) public view returns (uint256) {
+        return underlyingVault.maxMint(address(this));
+    }
+
+    function maxWithdraw(address owner) public view returns (uint256) {
+        uint256 maxWithdrawVault = underlyingVault.maxWithdraw(address(this));
+        uint256 assetsBalance = previewRedeem(balanceOf[owner]);
+        return maxWithdrawVault < assetsBalance ? maxWithdrawVault : assetsBalance;
+    }
+
+    function maxRedeem(address owner) public view returns (uint256) {
+        uint256 maxRedeemVault = underlyingVault.maxRedeem(address(this));
+        uint256 shareBalance = balanceOf[owner];
+        return maxRedeemVault < shareBalance ? maxRedeemVault : shareBalance;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              AD HOC LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function maxApproveUnderlyingVault() public {
+        asset.safeApprove(address(underlyingVault), type(uint256).max);
+    }
+
+    /// @dev In case something goes wrong with the underlying vault, we want the user to be able to claim the vault's
+    ///      tokens to interact directly with it
+    function redeemAllForUnderlyingVaultToken(uint256 shares, address receiver) public returns (uint256 assets) {
```

Comment:
This is something that I thought about after the Euler hack. In that case, the eTokens were left stuck inside the 4626 adapter vault
My concern is that if Euler decides to open up a claim contract were you could swap eTokens for their underlying (or something similar), then funds would be completely lost. So the idea is to add a way for users to redeem the proxy's tokens in exchange for the underlying vault's tokens

Reply:
Yeap. Great catch, def. something that we should have added on previos stuff. I want to communicate with the Euler team regarding our case, but still waiting for SOME update on their part tbh.